packages/vaadin-virtual-list/src/virtualizer-iron-list-adapter.js

```diff
@@ -0,0 +1,430 @@
+    }
+    this._previousMomentum = momentum;
+  }
+
+  /**
+   * Determines if the element has an ancestor that handles the scroll delta prior to this
+   *
+   * @private
+   */
+  _hasScrolledAncestor(el, deltaX, deltaY) {
+    if (el === this.scrollTarget || el === this.scrollTarget.getRootNode().host) {
+      return false;
+    } else if (
+      this._canScroll(el, deltaX, deltaY) &&
+      ['auto', 'scroll'].indexOf(getComputedStyle(el).overflow) !== -1
+    ) {
+      return true;
+    } else if (el !== this && el.parentElement) {
+      return this._hasScrolledAncestor(el.parentElement, deltaX, deltaY);
+    }
+  }
+
+  _canScroll(el, deltaX, deltaY) {
+    return (
+      (deltaY > 0 && el.scrollTop < el.scrollHeight - el.offsetHeight) ||
+      (deltaY < 0 && el.scrollTop > 0) ||
+      (deltaX > 0 && el.scrollLeft < el.scrollWidth - el.offsetWidth) ||
+      (deltaX < 0 && el.scrollLeft > 0)
+    );
+  }
+
+  /**
+   * @returns {Number|undefined} - The browser's default font-size in pixels
+   * @private
+   */
+  _getScrollLineHeight() {
+    const el = document.createElement('div');
+    el.style.fontSize = 'initial';
+    el.style.display = 'none';
+    document.body.appendChild(el);
+    const fontSize = window.getComputedStyle(el).fontSize;
+    document.body.removeChild(el);
+    return fontSize ? window.parseInt(fontSize) : undefined;
+  }
+
+  __getVisibleElements() {
+    return Array.from(this.elementsContainer.children).filter((element) => !element.hidden);
+  }
+
+  /** @private */
+  __reorderElements() {
+    if (this.__mouseDown) {
+      this.__pendingReorder = true;
+      return;
+    }
+    this.__pendingReorder = false;
+
+    const adjustedVirtualStart = this._virtualStart + (this._vidxOffset || 0);
+
+    // Which row to use as a target?
+    const visibleElements = this.__getVisibleElements();
+
+    const elementWithFocus = visibleElements.find(
+      (element) =>
+        element.contains(this.elementsContainer.getRootNode().activeElement) ||
+        element.contains(this.scrollTarget.getRootNode().activeElement)
+    );
+    const targetElement = elementWithFocus || visibleElements[0];
+    if (!targetElement) {
+      // All elements are hidden, don't reorder
+      return;
+    }
+
+    // Where the target row should be?
+    const targetPhysicalIndex = targetElement.__virtualIndex - adjustedVirtualStart;
+
+    // Reodrer the DOM elements to keep the target row at the target physical index
+    const delta = Array.from(visibleElements).indexOf(targetElement) - targetPhysicalIndex;
+    if (delta > 0) {
+      for (let i = 0; i < delta; i++) {
+        this.elementsContainer.appendChild(visibleElements[i]);
+      }
+    } else if (delta < 0) {
+      for (let i = visibleElements.length + delta; i < visibleElements.length; i++) {
+        this.elementsContainer.insertBefore(visibleElements[i], visibleElements[0]);
+      }
+    }
+  }
+
+  /** @private */
+  _adjustVirtualIndexOffset(delta) {
```

Comment:
This logic is invoked on "scroll" event and it updates the virtual index offset based on the new scroll position. The logic isn't new but was moved from [here](https://github.com/vaadin/web-components/blob/b163d3a8c200e6a9064e03b7ca1c4a34ebb4f3d9/packages/vaadin-grid/src/vaadin-grid-scroller.js#L268)